include/dali/core/partition.h

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_PARTITION_H_
+#define DALI_CORE_PARTITION_H_
+
+#include <type_traits>
+#include <algorithm>
+#include <tuple>
+#include <utility>
+#include "dali/core/util.h"
+
+namespace dali {
+namespace detail {
+
+template <typename Iterator, typename Predicate>
+auto multi_partition_impl(Iterator &&begin, Iterator &&end, Predicate &&pred) {
+  auto it = std::stable_partition(begin, end, pred);
+  return std::make_tuple(std::move(it));
+}
+
+template <typename Iterator, typename Predicate0, typename... Predicates>
+auto
+multi_partition_impl(Iterator &&begin, Iterator &&end, Predicate0 pred0, Predicates &&... preds) {
+  auto second_group_begin = multi_partition_impl(
+    begin,
+    end,
+    std::forward<Predicate0>(pred0));
+
+  auto other_groups = multi_partition_impl(
+    std::get<0>(second_group_begin),
+    end,
+    std::forward<Predicates>(preds)...);
+
+  return std::tuple_cat(second_group_begin, other_groups);
+}
+
+}  // namespace detail
+
+/**
+ * @brief Partitions a range of elements according to predicates `preds`
+ *
+ * @tparam Iterator     a random-access iterator
+ * @tparam Predicates   predicates callable on the elements of the collection
+ * @param begin         iterator pointing to the first element in the range
+ * @param end           iterator pointing to one-past the last element in the range
+ * @param preds         the predicates by which to group the elements
+ * @return A tuple of iterators pointing to the ends of regions satisfied by preds
+ *
+ * See the other overload for details.
+ */
+template <typename Iterator, typename... Predicates,
+         typename = std::tuple<
+            decltype(std::declval<Predicates>()(*std::declval<Iterator>()))...>>
```

Comment:
This idiom is a poor man's concept. It's becoming quite popular in C++ community for use where C++20 is not yet available.